packages/common/src/hooks/useHarmonyField.ts

```diff
@@ -0,0 +1,262 @@
+ * - Only show errors after a submission has been attempted
+ * - Apply transformations to the value on blur or change
+ * - Debounce validations and error displays to lighten the validation
+ *   workload or hide the errors while the user is still typing.
+ * - Allow overriding the Formik context's "validateOnChange"
+ *
+ * Most of the extra functionality is reserved for fields that users type into,
+ * eg. text, email, password inputs.
+ * Unlike the useField hook from Formik, this implementation relies much less
+ * on the event target in onChange and onBlur, instead using the passed in
+ * field name path. This makes usage with ReactNative more straightforward.
+ *
+ * @param propsOrFieldName the field's name, or the full props for the hook.
+ * @example
+ * const MyComponent = () => {
+ *    const [emailField, { error }] = useHarmonyField({ name: 'email', label: 'Email' })
+ *    return <TextInput {...field} error={!!error} helperText={error} />
+ * }
+ */
+export const useHarmonyField = <Value = any>(
+  propsOrFieldName: string | UseHarmonyFieldProps<Value>
+): [
+  FormikFieldInputProps<Value>,
+  FieldMetaProps<Value>,
+  FieldHelperProps<Value>
+] => {
+  const props =
+    typeof propsOrFieldName === 'string'
+      ? { name: propsOrFieldName }
+      : propsOrFieldName
+  const {
+    name,
+    type = 'text',
+    debouncedIdleMs = 0,
+    transformValueOnChange,
+    transformValueOnBlur,
+    showErrorBeforeSubmit = false,
+    validateFieldOnChange
+  } = props
+  const [field, meta, helpers] = useFormikField(props)
+  const { touched, error } = meta
+  const { setValue, setTouched } = helpers
+  const {
+    validateField,
+    validateOnBlur,
+    validateOnChange: validateFormOnChange,
+    submitCount
+  } = useFormikContext()
+
+  // Tracks whether the user is actively editing the field
+  const [isChanging, setIsChanging] = useState(false)
+
+  // Combine the form validateOnChange prop with this field's prop
+  const validateOnChange =
+    validateFieldOnChange === undefined
+      ? validateFormOnChange
+      : validateFieldOnChange
+
+  // Only show an error state if:
+  // - There's an error in validation, AND
+  // - The field has been touched (focused at least once), AND
+  // - The field can error before submission or the form has been submitted at least once, AND
+  // - The field is not currently being changed
+  const hasError = Boolean(
+    error &&
+      touched &&
+      (showErrorBeforeSubmit || submitCount > 0) &&
+      !isChanging
+  )
+
+  /**
+   * Sets the field to touched, validates it, and then sets isChanging to false.
+   * Essentially mimicking onBlur but ensuring validation finishes before
+   * allowing errors to show.
+   */
+  const validateAndSetIdle = useCallback(() => {
+    // We'll validate ourselves - pass false for validation
+    setTouched(true, false)
+    // Cast back to promise (the types are wrong)
+    const maybePromise = validateField(name) as Promise<any> | void
```

Comment:
So does this prevent form level validation from firing for this field, or is there a case that we get double validation?

Reply:
Good callout, sent me back into the rabbit hole a bit :joy:

What I found is basically everywhere we pass in a Zod schema as the validationSchema anyway, which means that there's really no difference between "field level" vs "form level" validation - the whole schema is validated no matter what. The goal of what I'm doing here is basically allowing us to wait for validation instead of doing it ahead of time. I just realized though that unlike the `field.onBlur` callback, `setTouched` returns the validation promise, so we don't need to split this out into two calls!

I'll make that change...!